File: cpp/se/ContactUpdateCommand.cpp
#include "se/ContactUpdateCommand.hpp"
#include "se/StandardObjectType.hpp"
#include "xml/XMLHelper.hpp"
#include "se/Status.hpp"
#include "se/IntPostalInfo.hpp"
#include "se/LocalPostalInfo.hpp"
#include "se/Disclose.hpp"

/*
 * The complete set of attributes of a contact which may be updated as per
 * RFC3733.
 */
ContactUpdateCommand::ContactUpdateCommand (const std::string &id,
                                            const std::string &pw,
                                            const std::vector<Status> *addStatuses,
                                            const std::vector<std::string> *remStatuses,
                                            const IntPostalInfo *newIntPostalInfo,
                                            const LocalPostalInfo *newLocPostalInfo,
                                            const std::string *newVoice,
                                            const std::string *newVoiceExt,
                                            const std::string *newFax,
                                            const std::string *newFaxExt,
                                            const std::string *newEmail,
                                            const Disclose *disclose)
    : UpdateCommand (StandardObjectType::CONTACT(), id)
{
    if (addStatuses)
    {
        const std::vector<Status> &stats = *addStatuses;
        
        DOMElement *add = xmlWriter->appendChild (objElement, "add");
        
        for (unsigned int i = 0; i < stats.size(); i++)
            xmlWriter->appendChild (add, "status", stats[i].getRationale(),
                                    "s", stats[i].toString());
    }
    
    if (remStatuses)
    {
        const std::vector<std::string> &stats = *remStatuses;
        
        DOMElement *rem = xmlWriter->appendChild (objElement, "rem");
        
        for (unsigned int i = 0; i < stats.size(); i++)
            xmlWriter->appendChild (rem, "status", "s", stats[i]);
    }
    
    if (pw == "" && 
        newIntPostalInfo == NULL && 
        newLocPostalInfo == NULL && 
        newVoice == NULL && 
        newFax == NULL &&
        newEmail == NULL && 
        disclose == NULL) return;

    DOMElement *chg = xmlWriter->appendChild (objElement, "chg");
    
    if (newIntPostalInfo)
        newIntPostalInfo->appendToElement (xmlWriter, chg);
    
    if (newLocPostalInfo)
        newLocPostalInfo->appendToElement (xmlWriter, chg);
    
    if (newVoice)
    {
        DOMElement *voice = xmlWriter->appendChild (chg, "voice");
        if (newVoiceExt)
			XMLHelper::setAttribute (voice, "x", *newVoiceExt);
        
		XMLHelper::setTextContent (voice, *newVoice);
    }
    
    if (newFax)
    {
        DOMElement *fax = xmlWriter->appendChild (chg, "fax");
        if (newFaxExt)
			XMLHelper::setAttribute (fax, "x", *newFaxExt);
    
		XMLHelper::setTextContent (fax, *newFax);
    }
    
    if (newEmail)
		XMLHelper::setTextContent
			(xmlWriter->appendChild (chg, "email"), *newEmail);
    
    if (pw != "")
		XMLHelper::setTextContent
			(xmlWriter->appendChild
            	(xmlWriter->appendChild
                	(chg,
                 	"authInfo"),
             	 "pw"),
			 pw);
    
    if (disclose)
        disclose->appendToElement (xmlWriter, chg);
}


                                            

File: src/main/java/com/ausregistry/jtoolkit2/se/ContactUpdateCommand.java
package com.ausregistry.jtoolkit2.se;

import com.ausregistry.jtoolkit2.ErrorPkg;

import org.w3c.dom.Element;

/**
 * Use this to request the update of a contact object provisioned in an EPP
 * Registry.  Instances of this class generate RFC5730 and RFC5733 compliant
 * contact update EPP command service elements via the toXML method.  The
 * response expected from a server should be handled by a Response object.
 *
 * @see com.ausregistry.jtoolkit2.se.Response
 */
public class ContactUpdateCommand extends UpdateCommand {
    private static final long serialVersionUID = -8110771660904052882L;

    /**
     * Minimal constructor for changing the authinfo of a contact.
     */
    public ContactUpdateCommand(String id, String pw) {
        this(id, pw, null, null, null, null, null, null, null, null,
                null, null);
    }

    /**
     * The complete set of attributes of a contact which may be updated as per
     * RFC5733.
     *
     * @throws IllegalArgumentException if {@code id} is {@code null}.
     */
    public ContactUpdateCommand(String id, String pw,
            Status[] addStatuses, String[] remStatuses,
            IntPostalInfo newIntPostalInfo, LocalPostalInfo newLocPostalInfo,
            String newVoice, String newVoiceExt, String newFax,
            String newFaxExt, String newEmail, Disclose disclose) {

        super(StandardObjectType.CONTACT, id);

        if (id == null) {
            throw new IllegalArgumentException(ErrorPkg.getMessage(
                        "se.contact.update.id.missing"));
        }

        if (addStatuses != null) {
            Element add = xmlWriter.appendChild(objElement, "add");

            for (Status status : addStatuses) {
                xmlWriter.appendChild(add, "status", status.getRationale(),
                        "s", status.toString());
            }
        }

        if (remStatuses != null) {
            Element rem = xmlWriter.appendChild(objElement, "rem");

            for (String status : remStatuses) {
                xmlWriter.appendChild(rem, "status", "s", status);
            }
        }

        if (pw == null && newIntPostalInfo == null && newLocPostalInfo == null
                && newVoice == null && newFax == null
                && newEmail == null && disclose == null) {
            return;
        }

        Element chg = xmlWriter.appendChild(objElement, "chg");

        if (newIntPostalInfo != null) {
            newIntPostalInfo.appendToElement(xmlWriter, chg);
        }

        if (newLocPostalInfo != null) {
            newLocPostalInfo.appendToElement(xmlWriter, chg);
        }

        if (newVoice != null) {
            Element voice = xmlWriter.appendChild(chg, "voice");
            if (newVoiceExt != null) {
                voice.setAttribute("x", newVoiceExt);
            }
            voice.setTextContent(newVoice);
        }

        if (newFax != null) {
            Element fax = xmlWriter.appendChild(chg, "fax");
            if (newFaxExt != null) {
                fax.setAttribute("x", newFaxExt);
            }
            fax.setTextContent(newFax);
        }

        if (newEmail != null) {
            xmlWriter.appendChild(chg, "email").setTextContent(newEmail);
        }

        if (pw != null) {
            xmlWriter.appendChild(
                    xmlWriter.appendChild(
                        chg,
                        "authInfo"),
                    "pw").setTextContent(pw);
        }

        if (disclose != null) {
            disclose.appendToElement(xmlWriter, chg);
        }
    }
}

